fix locale-sensitive comparison of UnmatchedArgument error strings

baronfel opened the pull request:
## Description

Closes #29543 
Closes #28975 

by fixing a helper method comparison that was intended to filter out unmatched token errors. Since errors in System.CommandLine parsing are strings, discovering the source of a particular string is hard. This method attempted to see if a 'masked' error string was entirely contained in the error string. This works for languages and grammars that put the argument name in single quotes at the end of the string, but for locales like `de-DE` both of these invariants were broken.

The new method breaks the format string at the  .NET format placeholders, then ensures that each of those literal portions exists in order in the final localized error message string.

This issue **was customer reported**.

## Customer Impact
This potentially impacts build, pack, test, restore, list references, and other CLI Commands that use this same call to check for unparsed arguments. Users that used these commands in locales with non-English arrangements of locale strings would not be able to execute those commands. One **workaround** would be to force the use of English (or similarly-structured) locales.

## Regression
**Yes**, but hidden under a couple other issues that we fixed in 7.0.101.

## Risk
**Low** - the new checking logic seems to be stable under different locales, and has tests covering the locales that users have reported failing.

## Testing
I locally tested this by using the DOTNET_CLI_UI_LANGUAGE environment variable to setup the triggering situation.
I added automated testing of different locales to validate the change.

## Follow up

This kind of string manipulation is very error-prone, so I've reached out to the System.CommandLine API design group to consider a more structured approach.

baronfel committed:
fix locale-sensitive comparison of UnmatchedArgument error strings

Evangelink commented on src/Cli/dotnet/ParseResultExtensions.cs:
```diff
@@ -34,18 +34,29 @@ public static void ShowHelp(this ParseResult parseResult)
             Parser.Instance.Parse(tokenList).Invoke();
         }
 
+        private static string StripPlaceholdersFromString(string input) {
```
Nitpick

```suggestion
        private static string StripPlaceholdersFromString(string input)
        {
```

Evangelink commented on src/Cli/dotnet/ParseResultExtensions.cs:
```diff
@@ -34,18 +34,29 @@ public static void ShowHelp(this ParseResult parseResult)
             {
-                var unrecognizedTokenErrors = parseResult.Errors.Where(error =>
-                    error.Message.Contains(Parser.Instance.Configuration.LocalizationResources.UnrecognizedCommandOrArgument(string.Empty).Replace("'", string.Empty)));
+                var unrecognizedTokenErrors = parseResult.Errors.Where(error => {
```
Nitpick:

```suggestion
                var unrecognizedTokenErrors = parseResult.Errors.Where(error => 
                {
```

Evangelink reviewed (commented):
Shall we add the test with `dotnet test` and non-US locale as part of this PR or a follow-up one?

baronfel committed:
Apply suggestions from code review

Co-authored-by: Amaury Levé <amaury.leve@gmail.com>

baronfel committed:
add test parser test covering locale setting

baronfel commented:
Just added a test that
* sets a locale
* parses a `dotnet test` command with an unmatched argument
* asserts that parse errors exist, and
* asserts that our throw helper doesn't throw (because the errors are UnmatchedToken errors

This test works with en-US and de-DE cultures, which seem to cover a number of the differences in culture strings here.

danmoseley commented on src/Cli/dotnet/ParseResultExtensions.cs:
```diff
@@ -34,18 +34,31 @@ public static void ShowHelp(this ParseResult parseResult)
+            // from an input string.
+            // This lets us do fair comparisons against strings that are translated, since the placeholders
+            // may be in various locations, and may have different quote characters in different languages.
+            return System.Text.RegularExpressions.Regex.Replace(input, @"['""](.*)['""]", string.Empty);
```
This is greedy .. it will change `Foo "bar" baz "biz" boo`
To `Foo boo` won't it?

You can change to `(.*?)` or `([^'""])*` if you don't want that.

baronfel replied:
Thanks! I'll do so. I want to remove each instance of the pattern.

baronfel committed:
use Dan's fix

dsplaisted commented on src/Cli/dotnet/ParseResultExtensions.cs:
```diff
@@ -34,18 +34,31 @@ public static void ShowHelp(this ParseResult parseResult)
+            // from an input string.
+            // This lets us do fair comparisons against strings that are translated, since the placeholders
+            // may be in various locations, and may have different quote characters in different languages.
+            return System.Text.RegularExpressions.Regex.Replace(input, @"['""](.*?)['""]", string.Empty);
```
I'm bad at regex, but doesn't this still assume that the quoting character is either a single or a double quote?

Would it be better to parse the format string into its components, and check that the error message contains the non-substituted components of the format string in the right order?

IE the string `Unrecognized command or argument '{0}'` has these components:

- `Unrecognized command or argument '` - literal
- '{0}' - substitution
- `'` - literal

So we would check that the error string started with the first component, and that the third component was found somewhere after the end of the first one.

That way I don't think we'd be dependent on localization or on the substituted arguments having quotes around them at all.

baronfel replied:
Yeah, that's a solid plan. I'll work on that today.

dsplaisted commented on src/Tests/dotnet.Tests/ParserTests/TestParserTests.cs:
```diff
@@ -0,0 +1,41 @@
+        [Theory]
+        public void TestParserCanGetArgumentFromDoubleDash(string culture)
+        {
+            System.Globalization.CultureInfo.CurrentCulture = new System.Globalization.CultureInfo(culture);
```
Tests are run in parallel, and this affects global state, so this may cause flakiness in our tests.

baronfel replied:
Yeah, I know - I couldn't find a way around this aside from reworking a lot of the internals of the CLI parser to use some kind of 'locale provider' abstraction instead. Definitely open to thoughts here.

MarcoRossignoli replied:
You can go out of proc like runtime tests https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.RemoteExecutor @ViktorHofer is there some guide on how to integrate?

ViktorHofer replied:
Unfortunately we don't have a README but you should just need a PackageReference pointing to the package and then invoke the API.

baronfel replied:
I suppose an alternative here would be to use the existing `DotnetTestCommand` wrappers that actually invoke the CLI on an actual project - but that is _a lot_ of additional work for what should just be a test of the parser.

baronfel replied:
I ended up using out-of-proc test executions against a known, simple test project to validate this.

MarcoRossignoli commented:
@baronfel  any news on this one?

baronfel committed:
update error-message-checking algorithm to extract literal portions of format string and check for ordered presence of those in the formatted message.

baronfel committed:
Use out of process tests instead.

Adds a helper method for setting culture to the test framework, driven from a shared reference to the hard-coded UI environment variable.

baronfel commented on src/Cli/Microsoft.DotNet.Cli.Utils/UILanguageOverride.cs:
```diff
@@ -4,11 +4,11 @@
 using System.Globalization;
 
-namespace Microsoft.DotNet.Cli
+namespace Microsoft.DotNet.Cli.Utils
```
I moved this file to Utils because a) much of the shared utility logic for the CLI was already in this project and b) this project already had internals visible to the test framework, where I wanted to add a `WithCulture` extension method.

baronfel commented:
I think we're in a good space now with the algorithm and the tests for this check.  I'll want to get another check from @dsplaisted to make sure he's comfortable with the implementation, but then we're ready to go.

A note about the timing of this fix: we are already beyond the merge window for the January servicing release of .NET 7 (which would be 7.0.102) due to the holiday season. As a result, this fix (when approved and merged) will only be available in the Februrary servicing release at earliest (so 7.0.103). This feels really bad, and isn't a great response time given the very fast time to reporting that users had, but it's the best I can do.

In the meantime, the best thing I can suggest for users that are hitting this error is to force the CLI to the English local for `dotnet test`. The easiest way to do that is to set the `DOTNET_CLI_UI_LANGUAGE` environment variable in your shell to `en-US` for the duration of the test command. e.g. in Bash you'd run `DOTNET_CLI_UI_LANGUAGE='en-us' dotnet test .....`. Full documentation for this environment variable can be found [here](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#dotnet_cli_ui_language).

MarcoRossignoli commented:
@baronfel thanks for the fix!

dsplaisted commented on src/Cli/dotnet/ParseResultExtensions.cs:
```diff
@@ -34,21 +35,51 @@ public static void ShowHelp(this ParseResult parseResult)
+            ///<summary>Splits a .NET format string by the format placeholders (the {N} parts) to get an array of the literal parts, to be used in message-checking</summary> 
+            static string[] DistinctFormatStringParts(string formatString)
+            {
+                return Regex.Split(formatString, @"{[0-9]+}"); // match the literal '{', followed by any of 0-9 one or more times, followed by the literal '}'
```
Are braces not special characters in .NET regexes?  Do they need to be escaped?

baronfel replied:
Correct, they are not special-cased:

![image](https://user-images.githubusercontent.com/573979/210655321-5188f2ea-e175-41d6-ae55-f2beb67a8d1f.png)

You can see here that the `Value` of the Capture is the literal brace-wrapped number.

dsplaisted commented on src/Cli/dotnet/ParseResultExtensions.cs:
```diff
@@ -34,21 +35,51 @@ public static void ShowHelp(this ParseResult parseResult)
+                    var foundIndex = error.IndexOf(part);
+                    if (foundIndex != -1)
+                    {
+                        error = error.Slice(foundIndex);
```
Should this be the following?

```suggestion
                        error = error.Slice(foundIndex + part.Length);
```

baronfel replied:
good eye!

baronfel committed:
Update src/Cli/dotnet/ParseResultExtensions.cs

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

marcpopMSFT commented:
@baronfel this was approved for February. Please ensure the checks get resolved and this gets merged by eow.

baronfel commented:
/azp run

baronfel commented:
Both of these check failures are due to the macos helix image expiration, which I think has already been handled in an earlier PR. Trying to kick these off again to see.
